### PR TITLE
cyrus: init at 3.4.3

### LINKED ISSUE #54012: Package request: cyrus-imap
https://cyrusimap.org

Module would be great, too. 
I might start working on this at some point, if my employer decides to self-host mail services.

### PR DESCRIPTION
**Pull request by pingiun**
###### Description of changes

Add cyrus imap server. Used compilation options from https://www.cyrusimap.org/imap/developer/compiling.html#setting-up-dependencies

For jmap support cyrus requires the unpackaged wslay, which I added as well.

Fixes #54012

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pingiun** [comment]
I believe cyrus should be able to work on macOS but libmd is labelled to not support darwin. I'm not sure why though

**Comment by pingiun** [comment]
Mentioning people from the original issue because you may be able to test this: @petabyteboy @teto @solson

**Review by teto — APPROVED** [review]
I've never used cyrus so I can't really tell if it works but I am glad it's available in nixpkgs. I am interested in the jmap support mostly

**Comment by pingiun** [comment]
The pidfile is hardcoded so i added the option to ./configure to place it at /run/cyrus-master.pid

**Comment by onny** [comment]
Thank you for your work, looks really good. Please put `wslay` in a separate commit

File: pkgs/servers/mail/cyrus/default.nix [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+  configureFlags = [
+    "--enable-autocreate"
+    "--enable-idled"
+    "--enable-murder"
+    "--enable-replication"
+    "--enable-unit-tests"
+    "--with-pidfile=/run/cyrus-master.pid"
+  ] ++ lib.optional (withHttp || withCalalarmd || withJMAP) "--enable-http"
+    ++ lib.optional withJMAP "--with-jmap"
```
Comment (sikmir) [inline pkgs/servers/mail/cyrus/default.nix:59]:
```suggestion
    ++ lib.withFeature withJMAP "jmap"
```

File: pkgs/servers/mail/cyrus/default.nix [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    ++ lib.optional withMySQL "--with-mysql"
+    ++ lib.optional withPgSQL "--with-pgsql"
+    ++ lib.optional withSQLite "--with-sqlite";
+
+  checkInputs = [ cunit ];
+  doCheck = true;
+
+  meta = with lib; {
+    homepage = "https://www.cyrusimap.org";
+    description = "Cyrus IMAP is an email, contacts and calendar server.";
```
Comment (sikmir) [inline pkgs/servers/mail/cyrus/default.nix:71]:
```suggestion
    description = "Cyrus IMAP is an email, contacts and calendar server";
```

**Comment by sikmir** [comment]
> I believe cyrus should be able to work on macOS but libmd is labelled to not support darwin. I'm not sure why though

`libmd` is already enabled on darwin efa78245a3d

**Comment by sikmir** [comment]
[3.4.4.](https://github.com/cyrusimap/cyrus-imapd/releases/tag/cyrus-imapd-3.4.4) is out.

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 170919` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>cyrus</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>wslay</li>
  </ul>
</details>

```
=======================================
   wslay 1.1.1: tests/test-suite.log
=======================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: main
==========

dyld[46531]: Library not loaded: '/nix/store/j82arv42js3p0s4m6mbmbzbb6d95lz54-wslay-1.1.1/lib/libwslay.0.dylib'
  Referenced from: '/private/tmp/nix-build-wslay-1.1.1.drv-0/wslay-1.1.1/tests/main'
  Reason: tried: '/nix/store/j82arv42js3p0s4m6mbmbzbb6d95lz54-wslay-1.1.1/lib/libwslay.0.dylib' (no such file), '/usr/local/lib/libwslay.0.dylib' (no>
FAIL main (exit status: 134)
```

File: pkgs/development/libraries/wslay/default.nix [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{ stdenv, lib, fetchurl, cunit }:
+
+stdenv.mkDerivation rec {
+  pname = "wslay";
+  version = "1.1.1";
+
+  src = fetchurl {
```
Comment (pacien) [inline pkgs/development/libraries/wslay/default.nix:7]: Why not `fetchFromGitHub`?

File: pkgs/servers/mail/cyrus/default.nix [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+, withMySQL ? false, libmysqlclient
+, withPgSQL ? false, postgresql
+, withSQLite ? true, sqlite
+}:
+
+stdenv.mkDerivation rec {
+  pname = "cyrus";
+  version = "3.4.3";
+
+  src = fetchurl {
```
Comment (pacien) [inline pkgs/servers/mail/cyrus/default.nix:18]: Why not `fetchFromGitHub`?

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -21559,6 +21561,8 @@ with pkgs;
     erlang = erlangR22;
   };
 
+  cyrus = callPackage ../servers/mail/cyrus { };
```
Comment (pacien) [inline pkgs/top-level/all-packages.nix:21564]: The project's website, source repo and doc are named Cyrus IMAP(d).
Even though it supports more than this protocol, we should probably stick
with the same name.

File: pkgs/servers/mail/cyrus/default.nix [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+, withXapian ? true, xapian
+, withCalalarmd ? true
+, withMySQL ? false, libmysqlclient
+, withPgSQL ? false, postgresql
+, withSQLite ? true, sqlite
+}:
+
+stdenv.mkDerivation rec {
+  pname = "cyrus";
+  version = "3.4.3";
```
Comment (pacien) [inline pkgs/servers/mail/cyrus/default.nix:16]: Version 3.8.0 is now out:
https://github.com/cyrusimap/cyrus-imapd/releases/tag/cyrus-imapd-3.8.0

**Review by pacien — COMMENTED** [review]
Thanks for packaging this.
I believe this is only a few minor changes away from being merged.

**Comment by pacien** [comment]
I just realised that this PR hasn't been updated for more than a year.

@pingiun: Do you still intend to work on this?
If not, I can take over this.
I have a package for 3.8.0 and a module draft as well.

**Comment by pingiun** [comment]
> I just realised that this PR hasn't been updated for more than a year.
> 
> @pingiun: Do you still intend to work on this? If not, I can take over this. I have a package for 3.8.0 and a module draft as well.

It's fine if you take over, I don't think I can finish this work soon

**Comment by Moraxyc** [comment]
It looks good. Are there any problems that block the merger of the PR?

**Comment by pingiun** [comment]
Closed in favour of #305538

**Pull request closed without merging**